### PR TITLE
[SPARK-16097][SQL] Encoders.tuple should handle null object correctly

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Although the top level input object can not be null, but when we use `Encoders.tuple` to combine 2 encoders, their input objects are not top level anymore and can be null. We should handle this case.
## How was this patch tested?

new test in DatasetSuite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @liancheng @clockfly

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -110,16 +110,25 @@ object ExpressionEncoder {
...
+      })
+
+      if (enc.flat) {
+        newSerializer.head
+      } else {
+        val struct = CreateStruct(newSerializer)
+        val nullCheck = Or(
+          IsNull(newInputObject),
+          Invoke(Literal.fromObject(None), "equals", BooleanType, newInputObject :: Nil))
+        If(nullCheck, Literal.create(null, struct.dataType), struct)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: This part is quite tricky, let's add comment here to explain why we need to substitute the input object and add the extra null check.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 2]
```diff
@@ -830,6 +830,13 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ds.dropDuplicates("_1", "_2"),
       ("a", 1), ("a", 2), ("b", 1))
   }
+
+  test("SPARK-16097: Encoders.tuple should handle null object correctly") {
+    val enc = Encoders.tuple(Encoders.tuple(Encoders.STRING, Encoders.STRING), Encoders.STRING)
+    val data = Seq((("a", "b"), "c"), (null, "d"))
+    val ds = spark.createDataset(data)(enc)
+    checkDataset(ds, (("a", "b"), "c"), (null, "d"))
+  }
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Shall we add the original outer join test case here?

**Comment by liancheng** [comment]
LGTM except for minor issues.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Actually that is not a valid test as the `ExpressionEncoder` is private and should not be used externally. But it does expose this bug and I added this test case which is valid externally.

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Is this equivalent with the original failed case? Seems not? For a outer join case, will we drop the null if `ExpressionEncoder` is not used?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 3]
```diff
@@ -110,16 +110,28 @@ object ExpressionEncoder {
...
-      case other => CreateStruct(other.serializer)
-    }.zipWithIndex.map { case (expr, index) =>
-      expr.transformUp {
-        case BoundReference(0, t, _) =>
-          Invoke(
-            BoundReference(0, ObjectType(cls), nullable = true),
-            s"_${index + 1}",
-            t)
+    val serializer = encoders.zipWithIndex.map { case (enc, index) =>
+      val originalInputObject = enc.serializer.head.collect { case b: BoundReference => b }.head
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Safe to call `head`?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Do we have any assumption at here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 4]
```diff
@@ -110,16 +110,28 @@ object ExpressionEncoder {
...
-    val serializer = encoders.map {
-      case e if e.flat => e.serializer.head
-      case other => CreateStruct(other.serializer)
-    }.zipWithIndex.map { case (expr, index) =>
-      expr.transformUp {
-        case BoundReference(0, t, _) =>
-          Invoke(
-            BoundReference(0, ObjectType(cls), nullable = true),
-            s"_${index + 1}",
-            t)
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Can you also comment on the line that caused the problem?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 1 above)
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: Also, let's put examples in the comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: this line: https://github.com/apache/spark/pull/13463/files#diff-87cabbe4d0c794f02523ecc1764955d0L88
we create struct directly and doesn't consider the null case

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: yea, https://github.com/apache/spark/pull/13807/files#diff-91c617f2464cea010922328f4cdbbda9R227

**Comment by liancheng** [comment]
Merging to master and branch-2.0.

**Pull request closed without merging**